### PR TITLE
[Lint] Reject methods shadowed by a later def in the same class

### PR DESCRIPTION
**Pull request by JasonKeyiL**
## Motivation

Python binds the last definition in a class body, so an earlier method with the same name never runs. It is usually a merge artifact — two branches each add the method, both land, and the survivor silently drops whatever the other one did. Nothing flags it, and the dead copy still reads as live code.

`HybridAttnBackend.forward` is the clearest example: #20812 proposed reverting #19369 in March 2026 precisely because it left a duplicate `forward`, that revert was auto-closed after 158 days, #32541 later added a third variant, and #34303 tried again in August and was closed unreviewed. Six months, three attempts, still there.

It also does not stay fixed. Scanning a week apart, `UnifiedMambaTokenToKVPoolAllocator.clear` had been repaired and `HybridLinearAttnBackend.get_indexer_metadata` had appeared. That is what makes this a gate rather than a one-off cleanup.

## Modifications

`scripts/lint/check_no_shadowed_defs.py`, wired into pre-commit next to the other AST checks. Only same-class collisions count, so a nested class keeps its own namespace; `@overload`, property accessors and dispatch registrations legitimately repeat a name and are skipped.

Removed the five shadowed definitions it finds. All are unreachable today, so behaviour is unchanged:

| class | method |
|---|---|
| `HybridLinearAttnBackend` | `get_indexer_metadata` |
| `DeepseekV4ForCausalLMDSpark` | `shared_experts_fusion_disable_reason` |
| `Gemma4ForCausalLM` | `get_embed_and_head` |
| `Gemma4ForConditionalGeneration` | `get_embed_and_head` |
| `TestSeparateReasoningExecution` | `tearDown` |

The first four keep the definition Python already bound, byte for byte — `ast.unparse` of the surviving def equals `ast.unparse` of the last def on `main`. The two `tearDown` bodies are the same three statements written twice, so that one keeps the documented copy rather than the terser one.

## Two things worth a maintainer's eye

**The dspark one had an NPU guard that was never running.** The shadowed copy returned a disable reason under `_is_npu`, saying ModelSlim weight loading cannot map shared experts into fused expert slots. The bound copy has no such branch, so on NPU that fusion is currently *not* disabled. Restoring it is a behaviour change, so it is not in this PR — but if that guard was meant to be live, it needs its own fix.

**A test is not testing what it names.** In `test_scripted_runtime_core.py`, `test_get_all_node_lock_refs_held_during_run_released_after` and `test_lock_refs_held_during_run_released_after` both call `self._script_lock_refs_held_then_released`, and that name is defined twice. The first script — the one that exercises `get_all_node_lock_refs()` — is the shadowed one, so it never runs and both tests execute the same script. Fixing that means renaming rather than deleting, and confirming it needs a GPU run, so I left the file excluded and flagged it here instead.

## Exclusions

Three pre-existing offenders are excluded, each with a note in the config so they are easy to retire:

- `qwen3_5_text.py` — being fixed in #38096
- `hybrid_attn_backend.py` — the history above; worth its own PR
- `test_scripted_runtime_core.py` — the coverage gap described above

## Test Plan

`scripts/lint/test_check_no_shadowed_defs.py`, 10 cases: shadowing detected for sync and async methods, every extra definition reported, and no false positives for a single def, same name in sibling classes, same name in a nested class, property setters, `typing.overload` stubs, module-level redefinition, or an unparsable file. The existing `check-lint-script-tests` hook picks it up automatically.

```
$ pre-commit run check-no-shadowed-defs --all-files
reject methods shadowed by a later def in the same class...............Passed

$ pre-commit run check-lint-script-tests --all-files
unit tests for lint checkers...........................................Passed
```

Full pre-commit chain is green on every changed file. Verified on macOS/arm64, Python 3.12; nothing here needs a GPU.

---
### CI States

Latest PR Test (Base): :x: [Run #34445759299](https://github.com/sgl-project/sglang/actions/runs/34445759299)
Latest PR Test (Extra): :x: [Run #34445759078](https://github.com/sgl-project/sglang/actions/runs/34445759078)
Latest PR Test (AMD ROCm 10): :x: [Run #34445759318](https://github.com/sgl-project/sglang/actions/runs/34445759318)

### CODE REVIEW

**Comment by JasonKeyiL** [comment]
Could someone add the `run-ci` label? `pr-gate` is failing on the missing label so nothing has run, and `/tag-run-ci-label` is a no-op for me.

Everything here is CPU-only: a new pre-commit check plus its unit tests, and five deletions of code Python never binds. For the four in `python/sglang/`, `ast.unparse` of the surviving definition is byte-identical to `ast.unparse` of the last definition on `main`, so there is no behaviour change to test for.

**Comment by haosenwang1018** [comment]
Heads-up while reviewing this: `python/sglang/srt/layers/attention/hybrid_attn_backend.py` is not in this diff, but `HybridAttnBackend.forward` is still defined twice there — line 161 (`73d272bddb6`, 2026-03-05, the one that accepts `mixed_qkv`/`a`/`b`) is shadowed by line 231 (`5b2e2750b55`, 2026-03-04, plain q/k/v). So the example your description calls out as the clearest one is still live on `main`.

Worth checking whether `check_no_shadowed_defs.py` would flag it as written; if it does, this PR needs that hunk too, otherwise the new gate fails on `main`. I put the one-file fix in #40840 — happy to close that and fold it in here instead if you would rather keep it all in one change.

**Comment by JasonKeyiL** [comment]
Thanks for the careful read, and for splitting #40840 out — I would keep it separate rather than fold it in.

On the gate itself: `hybrid_attn_backend.py` is already excluded in this PR's hook config, with a comment pointing at the history you cite, so the gate does pass on `main` as written. The red `lint` job was mine and unrelated — `check_no_shadowed_defs.py` landed without its executable bit while carrying a shebang, so `check-shebang-scripts-are-executable` failed. Fixed and pushed.

Your F811 note is the part that matters, and I think it argues against this PR as written. Measured on `main` at `954bb6804a`, over `python/ test/ scripts/`:

```
ruff --select=F811              27 findings
check_no_shadowed_defs.py        7 findings
```

F811 catches 6 of my 7. It misses only `test_scripted_runtime_core.py:277`, because the name is referenced between the two definitions and F811 is specifically about *unused* redefinitions.

It also catches 20 that my check does not — module-level defs, duplicate imports, class attributes — including both of the other cases I had found by hand and sent separately: `logits_processor.py:127` (#39707) and `deepseek_janus_pro.py:1095` (#39708). F811 would have found all three without a new file.

So adding `F811` to the existing ruff `--select` is strictly more coverage than my script and nothing new to maintain. The cost is that enabling it means resolving all 27 first, and some are noise rather than bugs — `DEFAULT_URL_FOR_TEST` re-exports and `diffusion_server` fixture imports would want per-file ignores.

I am happy to convert this PR into that: enable F811, fix the real ones, per-file-ignore the fixture and re-export cases. Or keep the narrow custom check if a 27-file blast radius is not wanted right now. @hnyls2002 @BBuf — which would you rather have?

One thing for #40840 while you are in there. "The linear-attention support at line 161 has never executed" is right, but I do not think it follows that linear attention is broken today. The surviving plain `forward` collects `mixed_qkv`/`a`/`b` into `**kwargs` and forwards them, and the child's `forward` (`hybrid_linear_attn_backend.py:1292`) declares `q`/`k`/`v` as `Optional[...] = None` alongside `mixed_qkv`/`a`/`b`, so the call still lands correctly. Framing it as dead code rather than a live breakage will hold up better under review.

**Comment by JasonKeyiL** [comment]
Pushed two fixes while the F811 question is open.

**The skip list was matching decorators by substring.** Anything whose text contained `register` was exempt, which is 121 `@register_custom_op`, 39 `@register_post_process`, 25 `@register_attention_backend` and more: 405 decorated functions skipped where only 44 actually rebind a name. A class with two same-named `@register_custom_op("...")` methods came back clean. Now matched on the decorator’s exact attribute, so rebinding only counts off an attribute (`@x.setter`, `@f.register`). Rescanning with the stricter predicate reports the same set as before, so nothing new needs excluding. Added the two tests that were missing — one collision that must be reported, one real `@handle.register` that must not.

**The red `lint` job was the executable bit**, not the gate: the checker carries a shebang like its siblings in `scripts/lint/` but landed `100644`, so `check-shebang-scripts-are-executable` failed. Fixed.

Also reworded the description — it called `HybridAttnBackend.forward` the clearest example without fixing it, which read oddly. It now points at #40840.
